kernel/src/process/program_loader/elf/load_elf.rs

```diff
@@ -222,31 +225,79 @@ pub fn map_segment_vmos(
             check_segment_align(program_header)?;
 
-            let map_at = relocated_range
-                .relocated_addr_of(program_header.virtual_addr as Vaddr)
-                .expect("Address not covered by `get_range_for_all_segments`");
+            // Reference: <https://elixir.bootlin.com/linux/v6.16.9/source/fs/binfmt_elf.c#L1200>
+            if program_header.flags.is_execute() {
+                // Update code segment range.
+                if program_header.virtual_addr < start_code {
+                    start_code = program_header.virtual_addr;
+                }
+                if end_code < program_header.virtual_addr + program_header.file_size {
+                    end_code = program_header.virtual_addr + program_header.file_size;
+                }
+            }
+            // Update data segment range.
+            if start_data < program_header.virtual_addr {
+                start_data = program_header.virtual_addr;
+            }
+            if end_data < program_header.virtual_addr + program_header.file_size {
+                end_data = program_header.virtual_addr + program_header.file_size;
+            }
+
+            let map_at = if elf_va_range.contains(&(program_header.virtual_addr as Vaddr)) {
+                map_range.start + (program_header.virtual_addr as Vaddr - elf_va_range.start)
+            } else {
+                panic!("Address not covered by `get_range_for_all_segments`");
+            };
 
             map_segment_vmo(program_header, elf_inode, vmar, map_at)?;
         }
     }
 
+    let relocated_range = RelocatedRange::new(
+        elf_va_range,
+        start_code as Vaddr..end_code as Vaddr,
+        start_data as Vaddr..end_data as Vaddr,
+        map_range.start,
+    )
+    .expect("Mapped range overflows");
+
     Ok(relocated_range)
 }
 
 /// A virtual range and its relocated address.
 pub struct RelocatedRange {
     original_range: Range<Vaddr>,
+    original_code_range: Range<Vaddr>,
+    original_data_range: Range<Vaddr>,
     relocated_start: Vaddr,
 }
 
 impl RelocatedRange {
     /// Creates a new `RelocatedRange`.
     ///
     /// If the relocated address overflows, it will return `None`.
-    pub fn new(original_range: Range<Vaddr>, relocated_start: Vaddr) -> Option<Self> {
+    pub fn new(
+        original_range: Range<Vaddr>,
+        original_code_range: Range<Vaddr>,
+        original_data_range: Range<Vaddr>,
+        relocated_start: Vaddr,
+    ) -> Option<Self> {
         relocated_start.checked_add(original_range.len())?;
+        // Ensure code range and data range are within original range.
+        if !original_range.contains(&original_code_range.start)
+            || !original_range.contains(&(original_code_range.end - 1))
+        {
+            return None;
+        }
+        if !original_range.contains(&original_data_range.start)
+            || !original_range.contains(&(original_data_range.end - 1))
+        {
+            return None;
+        }
```

Comment:
These checks also don't make sense. If you think they're prone to errors, there should be debug assertions. As I said before, though, we should remove the computation of `code_range` and `data_range` because I don't think they make a difference.